pkg/generators/models/testdata/cases/oneof_mapped_discriminator/expected/model_error.go

```diff
@@ -0,0 +1,141 @@
+// This file is auto-generated, DO NOT EDIT.
+//
+// Source:
+//     Title: Test oneOf discriminator support
+//     Version: 0.1.0
+package generatortest
+
+import (
+	"encoding/json"
+	"fmt"
+	validation "github.com/go-ozzo/ozzo-validation/v4"
+	"github.com/mitchellh/mapstructure"
+)
+
+type errorer interface {
+	GetKind() string
+	Validate() error
+}
+
+// Error is a oneOf type.
+type Error struct {
+	data errorer
+}
+
+// MarshalJSON implements the json.Marshaller interface
+func (m Error) MarshalJSON() ([]byte, error) {
+	return json.Marshal(m.data)
+}
+
+// UnmarshalJSON implements the json.Unmarshaller interface
+func (m *Error) UnmarshalJSON(bs []byte) error {
+	discriminator := struct {
+		Value string `json:"kind"`
+	}{}
+	err := json.Unmarshal(bs, &discriminator)
+	if err != nil {
+		return err
+	}
+
+	if discriminator.Value == "" {
+		return validation.Errors{
+			"kind": fmt.Errorf("can not be empty"),
+		}.Filter()
+	}
+
+	switch discriminator.Value {
+	case "auth":
+		data := GenericError{}
+		err := json.Unmarshal(bs, &data)
+		if err != nil {
+			return err
+		}
+		m.data = data
+	case "external":
+		data := ExternalError{}
+		err := json.Unmarshal(bs, &data)
+		if err != nil {
+			return err
+		}
+		m.data = data
+	case "field":
+		data := FieldError{}
+		err := json.Unmarshal(bs, &data)
+		if err != nil {
+			return err
+		}
+		m.data = data
+	case "generic":
+		data := GenericError{}
+		err := json.Unmarshal(bs, &data)
+		if err != nil {
+			return err
+		}
+		m.data = data
+	default:
+		return validation.Errors{
+			"kind": fmt.Errorf("invalid value"),
+		}.Filter()
+	}
+	return nil
+}
+
+// FromGenericError sets the Error data.
+func (m *Error) FromGenericError(data GenericError) {
+	m.data = data
+}
+
+// FromFieldError sets the Error data.
+func (m *Error) FromFieldError(data FieldError) {
+	m.data = data
+}
+
+// FromExternalError sets the Error data.
+func (m *Error) FromExternalError(data ExternalError) {
+	m.data = data
+}
+
+// As converts Error to a user defined structure.
+func (m Error) As(target interface{}) error {
+	return mapstructure.Decode(m.data, target)
+}
+
+// AsGenericError converts Error to a GenericError
+func (m Error) AsGenericError() (result GenericError, err error) {
```

Comment:
If we have a discriminiator we could return a error here if the user is requesting a type that is wrong. A `GetDiscriminatorValue()` function would also be handy from the user perspective

Reply:
I actually implemented exactly this last night :) i will push it in a few moments

Reply:
@trusch you can see the new implementation here https://github.com/contiamo/openapi-generator-go/pull/57/files#diff-93a8a6bbde7fe95808c421e882ab4a371ecec5480a123ee63ba18bd205a07874R159-R172

I avoided the `Get*` prefix because we generate `Get*` for each field, which means there is a small change for a name collision, instead, i use a `{{$modelName}}Discriminator`, in this case `ErrorDiscriminator` which we know we wont have any accidental name collisions with